File: BZEAD-iOS-main/BZEAD/Data/DestinationCountryRepository.swift
import Foundation

/// Mirrors web useDestinationCountry priority exactly.
final class DestinationCountryRepository {
    static let guestFallback = "United States"

    private let profileRepository = ProfileRepository()
    private let addressRepository = AddressRepository()

    func resolveCountry(session: BuyerSession, locationRepository: LocationRepository) async -> String {
        let candidates = await resolveCountryCandidates(session: session, locationRepository: locationRepository)
        return candidates.first ?? Self.guestFallback
    }

    func resolveCountryCandidates(session: BuyerSession, locationRepository: LocationRepository) async -> [String] {
        guard SupabaseConfig.isConfigured() else { return [Self.guestFallback] }

        let detected = await locationRepository.cachedLocation()?.country
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let shipping = (await CheckoutPreferencesRepository.readShippingCountry())
            .trimmingCharacters(in: .whitespacesAndNewlines)

        let profileCountryId = await profileRepository.fetchProfile(session: session)?.countryId
        let profileCountry = profileCountryId.flatMap { await fetchCountryName(countryId: $0) }?
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""

        let addresses = await addressRepository.fetchAddresses(session: session)
        let defaultAddress = addresses.first(where: { $0.isDefault }) ?? addresses.first
        let addressCountry = defaultAddress?.country.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""

        let ordered = [shipping, profileCountry, detected, addressCountry]
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }

        var distinct: [String] = []
        for value in ordered where !distinct.contains(value) {
            distinct.append(value)
        }
        return distinct.isEmpty ? [Self.guestFallback] : distinct
    }

    func resolveLocationLabel(session: BuyerSession, locationRepository: LocationRepository) async -> String {
        let pricingCountry = await resolveCountry(session: session, locationRepository: locationRepository)
        let addresses = await addressRepository.fetchAddresses(session: session)
        let defaultAddress = addresses.first(where: { $0.isDefault }) ?? addresses.first
        if let defaultAddress {
            let label = [defaultAddress.city, defaultAddress.state, defaultAddress.country]
                .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
                .filter { !$0.isEmpty }
                .joined(separator: ", ")
            if !label.isEmpty { return label }
        }

        if let cached = await locationRepository.cachedLocation()?.label(),
           !cached.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return cached
        }

        return pricingCountry
    }

    private func fetchCountryName(countryId: String) async -> String? {
        let http = SupabaseHTTP.shared
        let url = "\(SupabaseConfig.url)/rest/v1/countries" +
            "?id=eq.\(countryId)" +
            "&select=country_name,short_code,country_code,iso2" +
            "&limit=1"
        do {
            let (data, response) = try await http.getAllowingErrorStatus(url, headers: http.anonHeaders())
            guard http.isSuccess(response) else { return nil }
            let row = try http.decode([CountryNameRow].self, from: data).first
            return row?.countryName ?? row?.shortCode ?? row?.countryCode ?? row?.iso2
        } catch {
            return nil
        }
    }

    private struct CountryNameRow: Codable {
        let countryName: String?
        let shortCode: String?
        let countryCode: String?
        let iso2: String?
    }
}

File: BZEAD-APK-main/app/src/main/java/com/bzead/apk/data/DestinationCountryRepository.kt
package com.bzead.apk.data

import com.bzead.apk.data.model.BuyerSession
import io.ktor.client.call.body
import io.ktor.client.request.get
import com.bzead.apk.data.SupabaseHttp.anonHeaders
import io.ktor.http.isSuccess
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Mirrors web [useDestinationCountry] priority exactly.
 *
 * Logged-in: profile country (DB) → default address (DB). No local prefs / GPS —
 * the profile country is required at signup, so it is the authoritative source
 * and a locally-saved shipping country can never override it.
 * Guest: detected GPS → United States fallback.
 */
class DestinationCountryRepository {
    companion object {
        const val GUEST_FALLBACK = "United States"
    }

    /** Primary country used for pricing (same as web `selectedCountry`). */
    suspend fun resolveCountry(
        session: BuyerSession,
        locationRepository: LocationRepository,
    ): String {
        return resolveCountryCandidates(session, locationRepository).firstOrNull()
            ?: GUEST_FALLBACK
    }

    /**
     * Ordered list of countries to try for pricing RPCs (web resolution order).
     */
    suspend fun resolveCountryCandidates(
        session: BuyerSession,
        locationRepository: LocationRepository,
    ): List<String> {
        if (!SupabaseConfig.isConfigured()) return listOf(GUEST_FALLBACK)

        // Guest (no signed-in user): fall back to detected GPS, then US.
        if (session.userId.isBlank()) {
            val detected = locationRepository.cachedLocation()?.country?.trim().orEmpty()
            return listOf(detected)
                .filter { it.isNotBlank() }
                .distinct()
                .ifEmpty { listOf(GUEST_FALLBACK) }
        }

        // Signed-in users: DB only. Profile country (required at signup) wins, then
        // the DB default address. Local prefs / GPS are never consulted so a saved
        // foreign shipping country cannot override the user's profile country.
        val profileCountry = ProfileRepository().fetchProfile(session)?.countryId
            ?.let { fetchCountryName(it)?.trim().orEmpty() }
            .orEmpty()

        val addresses = AddressRepository().fetchAddresses(session)
        val addressCountry = addresses.firstOrNull { it.isDefault }?.country?.trim().orEmpty()
            .ifBlank { addresses.firstOrNull()?.country?.trim().orEmpty() }

        val ordered = listOf(profileCountry, addressCountry)
            .map { it.trim() }
            .filter { it.isNotBlank() }
            .distinct()

        return if (ordered.isEmpty()) listOf(GUEST_FALLBACK) else ordered
    }

    suspend fun resolveLocationLabel(
        session: BuyerSession,
        locationRepository: LocationRepository,
    ): String {
        val pricingCountry = resolveCountry(session, locationRepository)
        val addressRepo = AddressRepository()
        val addresses = addressRepo.fetchAddresses(session)
        val defaultAddress = addresses.firstOrNull { it.isDefault } ?: addresses.firstOrNull()
        if (defaultAddress != null) {
            val label = listOfNotNull(
                defaultAddress.city.takeIf { it.isNotBlank() },
                defaultAddress.state.takeIf { it.isNotBlank() },
                defaultAddress.country.takeIf { it.isNotBlank() },
            ).joinToString(", ")
            if (label.isNotBlank()) return label
        }

        locationRepository.cachedLocation()?.label()?.takeIf { it.isNotBlank() }?.let { return it }

        return pricingCountry
    }

    /** Signed-in user's saved profile country (id + display name), or null. */
    suspend fun fetchProfileCountry(session: BuyerSession): ProfileCountry? {
        if (!SupabaseConfig.isConfigured()) return null
        val countryId = ProfileRepository().fetchProfile(session)?.countryId?.trim().orEmpty()
        if (countryId.isBlank()) return null
        val name = fetchCountryName(countryId)?.trim().orEmpty()
        if (name.isBlank()) return null
        return ProfileCountry(countryId, name)
    }

    /**
     * Resolve a detected location (ISO2 code preferred, name fallback) to a countries
     * row so it can be compared against / written to the profile. Null if unknown.
     */
    suspend fun lookupCountryByLocation(country: String, iso2: String): DetectedCountry? {
        if (!SupabaseConfig.isConfigured()) return null

        val codeToken = iso2.trim().uppercase()
        val nameToken = country.trim()
        if (codeToken.isBlank() && nameToken.isBlank()) return null

        val filter = if (codeToken.isNotBlank()) {
            "iso2=eq.$codeToken"
        } else {
            "country_name=ilike.${java.net.URLEncoder.encode(nameToken, "UTF-8")}"
        }

        val response = SupabaseHttp.client.get(
            "${SupabaseConfig.url}/rest/v1/countries" +
                "?$filter" +
                "&select=id,country_name,currency_code" +
                "&limit=1",
        ) { anonHeaders() }
        if (!response.status.isSuccess()) return null
        val row = response.body<List<CountryLookupRow>>().firstOrNull() ?: return null
        val id = row.id?.trim().orEmpty()
        if (id.isBlank()) return null
        return DetectedCountry(
            id = id,
            name = row.countryName?.trim().orEmpty(),
            currencyCode = row.currencyCode?.trim()?.uppercase().orEmpty(),
        )
    }

    private suspend fun fetchCountryName(countryId: String): String? {
        val response = SupabaseHttp.client.get(
            "${SupabaseConfig.url}/rest/v1/countries" +
                "?id=eq.$countryId" +
                "&select=country_name,short_code,country_code,iso2" +
                "&limit=1",
        ) { anonHeaders() }
        if (!response.status.isSuccess()) return null
        val row = response.body<List<CountryNameRow>>().firstOrNull() ?: return null
        return row.countryName ?: row.shortCode ?: row.countryCode ?: row.iso2
    }

    data class ProfileCountry(val id: String, val name: String)

    data class DetectedCountry(val id: String, val name: String, val currencyCode: String)

    @Serializable
    private data class CountryLookupRow(
        val id: String? = null,
        @SerialName("country_name") val countryName: String? = null,
        @SerialName("currency_code") val currencyCode: String? = null,
    )

    @Serializable
    private data class CountryNameRow(
        @SerialName("country_name") val countryName: String? = null,
        @SerialName("short_code") val shortCode: String? = null,
        @SerialName("country_code") val countryCode: String? = null,
        val iso2: String? = null,
    )
}
